### PR TITLE
add file mounted crl support for client tls settings

### PR DESCRIPTION
**Pull request by ramaraochavali**
Followup #49346 for https://github.com/istio/api/pull/3052

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@kfaseela FYI

**Comment by kfaseela** [comment]
@leosarra could you also please review?

**Review by kfaseela — COMMENTED** [review]
What happens if both credentialName and CaCrl are specified? Do we have validations in place?

**Comment by leosarra** [comment]
> What happens if both credentialName and CaCrl are specified? Do we have validations in place?

I presume that the following check will be triggered in the validation but Rama could you please confirm if this is correct?
```
if tls.CredentialName != "" {
			v = appendValidation(v, fmt.Errorf("CRL is not supported with credentialName. CRL has to be specified in the credential"))
}
```
Other than that, LGTM

**Comment by ramaraochavali** [comment]
Thanks @kfaseela for the review. We need to fix validation code at one place. Fixed now. PTAL

**Comment by leosarra** [comment]
Do we need to add a check on caCrl also here?

https://github.com/istio/istio/blob/c45af9d39bd580390da19e962a3de53200a90230/pkg/config/validation/validation.go#L628

File: pilot/pkg/networking/core/v1alpha3/cluster_tls.go [hunk 1]
```diff
@@ -212,12 +212,21 @@ func (cb *ClusterBuilder) buildUpstreamClusterTLSContext(opts *buildClusterOpts,
 			if !res.IsRootCertificate() || tls.GetInsecureSkipVerify().GetValue() {
 				tlsContext.CommonTlsContext.ValidationContextType = &tlsv3.CommonTlsContext_ValidationContext{}
 			} else {
+				defaultValidationContext := &tlsv3.CertificateValidationContext{MatchSubjectAltNames: util.StringToExactMatch(tls.SubjectAltNames)}
+				if tls.GetCaCrl() != "" {
+					defaultValidationContext.Crl = &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: tls.GetCaCrl(),
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster_tls.go:219]: how do user put this files to a pod on the fly

**Review by hzxuzhonghu — COMMENTED** [review]
does not implement Gateway tls?

**Comment by ramaraochavali** [comment]
> Do we need to add a check on caCrl also here?

I think that is for gateways. I will take a look and fix it in separate PR

**Comment by ramaraochavali** [comment]
> does not implement Gateway tls?

https://github.com/istio/istio/pull/49346 - already merged for gateways

File: pilot/pkg/networking/core/v1alpha3/cluster_tls.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster_tls.go:219]: We build tooling to place certs/crls in the pod and replace it when they change.

**Comment by kfaseela** [comment]
> > Do we need to add a check on caCrl also here?
> 
> I think that is for gateways. I will take a look and fix it in separate PR

Please do fix this for gateways.

**Comment by kfaseela** [comment]
> > > Do we need to add a check on caCrl also here?
> > 
> > I think that is for gateways. I will take a look and fix it in separate PR
> 
> Please do fix this for gateways.

I see that it is already fixed and merged. Thanks!

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**